plugins/steelseries/fu-steelseries-fizz-tunnel.c

```diff
@@ -0,0 +1,395 @@
+	 *
+	 * LEVEL: 7 least significant bit value of the battery. Values are between 2-21, to get %
+	 * you can do (LEVEL - 1) * 5
+	 */
+	fu_device_set_battery_level(device,
+				    ((level & STEELSERIES_FIZZ_BATTERY_LEVEL_STATUS_BITS) - 1U) *
+					5U);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_steelseries_fizz_tunnel_wait_for_reconnect_cb(FuDevice *device,
+						 gpointer user_data,
+						 GError **error)
+{
+	FuDevice *parent = fu_device_get_parent(device);
+	guint8 status;
+
+	if (!fu_steelseries_fizz_connection_status(parent, &status, error)) {
+		g_prefix_error(error, "failed to get connection status: ");
+		return FALSE;
+	}
+	g_debug("ConnectionStatus: %u", status);
+	if (status == STEELSERIES_FIZZ_CONNECTION_STATUS_NOT_CONNECTED) {
+		g_set_error(error, FWUPD_ERROR, FWUPD_ERROR_NOT_FOUND, "device is unreachable");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_steelseries_fizz_tunnel_wait_for_reconnect(FuDevice *device, guint delay, GError **error)
+{
+	return fu_device_retry_full(device,
+				    fu_steelseries_fizz_tunnel_wait_for_reconnect_cb,
+				    delay / 1000,
+				    1000,
+				    NULL,
+				    error);
+}
+
+static gboolean
+fu_steelseries_fizz_tunnel_attach(FuDevice *device, FuProgress *progress, GError **error)
+{
+	FuDevice *parent = fu_device_get_parent(device);
+	guint remove_delay = fu_device_get_remove_delay(device);
+	g_autoptr(GError) error_local = NULL;
+
+	if (!fu_steelseries_fizz_reset(parent,
+				       TRUE,
+				       STEELSERIES_FIZZ_RESET_MODE_NORMAL,
+				       &error_local))
+		g_warning("failed to reset: %s", error_local->message);
+	/* wait for receiver to reset the connection status to 0 */
+	g_usleep(2000000); /* 2 s */
+	remove_delay -= 2000;
+
+	if (!fu_steelseries_fizz_tunnel_wait_for_reconnect(device, remove_delay, error)) {
+		g_prefix_error(error, "device %s did not come back", fu_device_get_id(device));
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_steelseries_fizz_tunnel_probe(FuDevice *device, GError **error)
+{
+	FuDevice *parent = fu_device_get_parent(device);
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(parent));
+	guint16 release;
+
+	/* set the version if the release has been set */
+	release = g_usb_device_get_release(usb_device);
+	if (release != 0x0 &&
+	    fu_device_get_version_format(device) == FWUPD_VERSION_FORMAT_UNKNOWN) {
+		g_autofree gchar *version = NULL;
+		version = fu_common_version_from_uint16(release, FWUPD_VERSION_FORMAT_BCD);
+		fu_device_set_version_format(device, FWUPD_VERSION_FORMAT_BCD);
+		fu_device_set_version(device, version);
+	}
+
+	/* add GUIDs in order of priority */
+	fu_device_add_instance_str(device, "STEELSERIES", "FIZZ");
+	fu_device_add_instance_u16(device, "VID", g_usb_device_get_vid(usb_device));
+	fu_device_add_instance_u16(device, "PID", g_usb_device_get_pid(usb_device));
+	fu_device_add_instance_u16(device, "REV", release);
```

Comment:
if != 0 perhaps?

Reply:
maybe; I have taken the code from [fu-usb-device.c](https://github.com/fwupd/fwupd/blob/1.8.1/libfwupdplugin/fu-usb-device.c#L450-L466); do you want I make the change, in both places?

Reply:
also, do you think `USB\STEELSERIES_FIZZ&VID_1038&PID_1838` is good `instance_id` name for child device over the 2.4G? the same (child) device over USB would have been `USB\VID_1038&PID_183A` (i.e. -- `SLEELSERIES_FIZZ` and `s,1838`,PID_183A,`).

Reply:
> do you want I make the change, in both places

No; in some cases we want REV_00 for firmware matching. Maybe don't change anything here.

Reply:
> , do you think USB\STEELSERIES_FIZZ&VID_1038&PID_1838

It's not really a USB subsystem at this point, is it? Maybe `FIZZ\VID_1038&PID_1838` is better?

Reply:
or `USB\VID_1038&PID_1838&MODE_FIZZ` or something.

Reply:
Right, it is not an USB, so better to start with `STEELSERIES` I guess.

What do you think of `STEELSERIES\FIZZ_TUNNEL&VID_1038&PID_1838`?

Then, I would like to introduce `STEELSERIES\FIZZ_HID&VID_0111&PID_1838` for detected mouse that are connected via bluetooth.

`_TUNNEL` for 2.4G and `_HID` for BT.

BTW, I would need you help on that one, I am not able to list such device for now. but this is out of scope for now.

Reply:
The instance ID convention is `KEY_VALUE` so you really want something like `STEELSERIES\TRANSPORT_FIZZHID&VID_0111&PID_1838` or `STEELSERIES\VID_0111&PID_1838&PROTOCOL_FIZZ_TUNNEL` or something like that.

Reply:
Hum, in my example, you mean `FIZZ_` is not an appropriate key and I should pickup `TRANSPORT_` or `PROTOCOL_` instead? with `_FIZZ<HID|TUNNEL>` as value?

`FIZZ_TUNNEL` and `FIZZ_HID` are not good names?

Reply:
Agree, fizz isn't really a key.

Reply:
Also, maybe stick the protocol last. Then we can match vid PID for quirks

Reply:
`STEELSERIES\VID_1038&PID_1838&PROTOCOL_FIZZ_TUNNEL`